src/js/utils/AsyncExpiringMap.ts

```diff
@@ -0,0 +1,163 @@
+/**
+ * A Map that automatically removes entries after their TTL has expired.
+ *
+ * The map is Promise-aware, meaning it will start TTL countdown only after the promise has resolved.
+ */
+export class AsyncExpiringMap<K, V> {
+  private _ttl: number;
+  private _cleanupIntervalMs: number;
+  private _map: Map<K, { value: V | undefined; expiresAt: number | null; promise: PromiseLike<V> | null }>;
+  private _cleanupInterval: ReturnType<typeof setInterval>;
+
+  public constructor({
+    cleanupInterval = 5_000,
+    ttl = 2_000,
+  }: {
+    cleanupInterval?: number;
+    ttl?: number;
+  } = {}) {
+    this._ttl = ttl;
+    this._map = new Map();
+    this._cleanupIntervalMs = cleanupInterval;
+    this.startCleanup();
+  }
+
+  /**
+   * Set a key-value pair.
+   */
+  public set(key: K, promise: PromiseLike<V> | V): void {
+    if (!this._cleanupInterval) {
+      this.startCleanup();
+    }
+
+    if (typeof promise !== 'object' || !('then' in promise)) {
+      this._map.set(key, { value: promise, expiresAt: Date.now() + this._ttl, promise: null });
+      return;
+    }
+
+    const entry: { value: V | undefined; expiresAt: number | null; promise: PromiseLike<V> | null } = {
+      value: undefined,
+      expiresAt: null,
+      promise,
+    };
+    this._map.set(key, entry);
+
+    promise.then(
+      value => {
+        entry.value = value;
+        entry.expiresAt = Date.now() + this._ttl;
+        entry.promise = null;
+      },
+      () => {
```

Comment:
Q: Should we log when an error occurs ?

Reply:
No, it's responsibility of the user of the Map to handle the promise rejection.